README revamp (#567)
VirtusLab/git-machete-intellij-plugin: README.md

![](docs/open_git_machete.gif)

> ### Branch graph

Author: This section was moved from `Edit machete file` section and slightly extended.

> For each branch, Git Machete indicates a relation between this branch and each of its child branches.

Reviewer: Maybe just
```suggestion
For each branch, Git Machete indicates the relation to each of its child branches.
```

If the edge between them is green that means the child branch is in sync with its parent branch &mdash; in other words, there are no commits in the parent branch that don't belong to the child.
But if there are some commits in the parent branch that are **not** reachable from the child, then the edge is red &mdash; you need to [rebase](#rebase) child branch onto the parent.
Edge can also be gray and yellow &mdash; that means respectively the branch was merged to the parent, or a fork point can't be determined automatically.

![](docs/sample_graph.png)

> As we can see in example above `hotfix/add-trigger` is in sync with `master`.

Reviewer: ```suggestion
As we can see in the example above, `hotfix/add-trigger` is in sync with `master`.
```

`call-ws` is **not** in sync with `develop` and `drop-constraint` is **not** in sync with `call-ws`.
`build-chain` was merged into `develop`.

> 

Reviewer: I'd still add some link to the Edit section so that it's clear that the branch layout can be arbitrarily modified by a user (i.e. is not fixed in stone)

![](docs/open_git_machete.gif)

### Branch graph

For each branch, Git Machete indicates the relation to each of its child branches.
If the edge between them is **green** that means the child branch is in sync with its parent branch &mdash; in other words, there are no commits in the parent branch that don't belong to the child.
But if there are some commits in the parent branch that are **not** reachable from the child, then the edge is **red** &mdash; you need to [rebase](#rebase) child branch onto the parent.

> Edge can also be **gray** or **yellow** &mdash; that means respectively the branch was merged to the parent, or a fork point can't be determined automatically

Reviewer: Maybe just split this sentence into two - one for gray, one for yellow (occurrence of `mdash`/`respectively`/`that == the latter` in a single sentence is a good indicator for split)

> (see [Override fork point section](#override-fork-point) to learn how to fix that).

Reviewer: Nit: for clarity:
```suggestion
(see [Override fork point section](#override-fork-point) to learn how to fix the latter).
```

![](docs/sample_graph.png)

As we can see in the example, above `hotfix/add-trigger` is in sync with `master`.
`call-ws` is **not** in sync with `develop` and `drop-constraint` is **not** in sync with `call-ws`.
`build-chain` was merged into `develop`.

> Note that the branch layout can be easily arbitrarily modified (see [Edit machete file section](#edit-machete-file) for details).

Reviewer: Nit: two `-ly` adverbs in a row don't sound well here, maybe
```suggestion
Note that the branch layout can be arbitrarily modified with ease (see [Edit machete file section](#edit-machete-file) for details).
```

It'll be located under `.git/machete` path in your repository.

This branch layout can be automatically discovered based on the state of your git repository by the `Discover Branch Layout` action.
It constructs a layout from around 10 most recently used branches.
**This action is automatically invoked in case of an empty or nonexistent `machete` file,**
but you can also run it any time from IntelliJ's `Search Everywhere` (double Shift) by typing `Discover Branch Layout`.

![](docs/discover.gif)

### Edit machete file

The `machete` file can be also edited manually. Let's look at the structure of this file based on the example below:
```
develop
    allow-ownership-link PR #123
        build-chain
    call-ws PR #124
master
    hotfix/add-trigger
```
`develop`, `allow-ownership-link`, `build-chain`, `call-ws`, `master` and `hotfix/add-trigger` are branch names.
Two of them, `allow-ownership-link` and `call-ws`, have a custom annotation &mdash;
it's an arbitrary description displayed next to the given branch (in this case, pull request numbers).

The relation between these branches is determined by indentations &mdash; here single indent is 4 spaces, but a tab can be used as well.

> In the example above, branches `allow-ownership-link` and `call-ws` are children of `develop`, while `build-chain` is a child of `allow-ownership-link`.
> `master`, in turn, is the parent of `hotfix/add-trigger`.
> `develop` and `master` are root branches.

Reviewer: For better readability:
```suggestion
In the example above, branches `allow-ownership-link` and `call-ws` are children of `develop`, while `build-chain` is a child of `allow-ownership-link`. <br/>
`master`, in turn, is the parent of `hotfix/add-trigger`. <br/>
`develop` and `master` are root branches.
```
